src/nyx/features/gabor.cpp

```diff
@@ -89,6 +91,216 @@ void GaborFeature::calculate (LR& r)
     }
 }
 
+#ifdef USE_GPU
+void GaborFeature::calculate_gpu (LR& r)
+{
+    // Skip calculation in case of bad data
+    if ((int)r.fvals[MIN][0] == (int)r.fvals[MAX][0])
+    {
+        fvals.resize (GaborFeature::num_features, 0);   // 'fvals' will then be picked up by save_values()
+        return;
+    }
+
+    const ImageMatrix& Im0 = r.aux_image_matrix;
+
+    double GRAYthr;
+    /* parameters set up in complience with the paper */
+    double gamma = 0.5, sig2lam = 0.56;
+    int n = 38;
+    double f0[7] = { 1, 2, 3, 4, 5, 6, 7 };       // frequencies for several HP Gabor filters
+    double f0LP = 0.1;     // frequencies for one LP Gabor filter
+    double theta = 3.14159265 / 2;
+    int ii;
+    unsigned long originalScore = 0;
+
+    readOnlyPixels im0_plane = Im0.ReadablePixels();
+
+    if (fvals.size() != GaborFeature::num_features)
+        fvals.resize (GaborFeature::num_features, 0.0);
+    
+    // Temp buffers
+
+    // --1
+    ImageMatrix e2img;
+    e2img.allocate (Im0.width, Im0.height);
+
+    //std::cout << "width: " << Im0.width << std::endl;
+    //std::cout << "height: " << Im0.height << std::endl;
+
+    // --2
+    std::vector<double> auxC ((Im0.width + n - 1) * (Im0.height + n - 1) * 2);
+
+    // --3
+    std::vector<double> auxG (n * n * 2);
+
+    // compute the original score before Gabor
+    GaborEnergyGPU (Im0, e2img.writable_data_ptr(), auxC.data(), auxG.data(), f0LP, sig2lam, gamma, theta, n);
+    readOnlyPixels pix_plane = e2img.ReadablePixels();
+    // N.B.: for the base of the ratios, the threshold is 0.4 of max energy,
+    // while the comparison thresholds are Otsu.
+
+    Moments2 local_stats;
+    e2img.GetStats(local_stats);
+    double max_val = local_stats.max__();
+
+    //
+    //originalScore = (pix_plane.array() > max_val * 0.4).count();
+    //
+    int cnt = 0;
+    double cmp_a = max_val * 0.4;
+    for (auto a : pix_plane)
+        if (double(a) > cmp_a)
+            cnt++;
+    originalScore = cnt;
+
+    for (ii = 0; ii < GaborFeature::num_features; ii++)
+    {
+        unsigned long afterGaborScore = 0;
+        writeablePixels e2_pix_plane = e2img.WriteablePixels();
+        GaborEnergyGPU (Im0, e2_pix_plane.data(), auxC.data(), auxG.data(), f0[ii], sig2lam, gamma, theta, n);
+
+        //
+        //Moments2 local_stats2;
+        //e2img.GetStats(local_stats2);
+        //double max_val2 = local_stats2.max__();
+        //
+        //e2_pix_plane.array() = (e2_pix_plane.array() / max_val2).unaryExpr(Moments2func(e2img.stats));
+        //

```

Comment:
Same comments as Andiy. IMO, removing dead and commented out codes increase readability. I am also guilty of that at times :-(

Reply:
Code has been cleaned up.